File: Python/0064_minimum-path-sum.py
"""=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 * 64. Minimum Path Sum
 *
 * Medium
 *
 * Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right,
 * which minimizes the sum of all numbers along its path.
 *
 * Note: You can only move either down or right at any point in time.
 *
 * Example 1:
 * Input: grid = [[1,3,1],[1,5,1],[4,2,1]]
 * Output: 7
 * Explanation: Because the path 1 → 3 → 1 → 1 → 1 minimizes the sum.
 *
 * Example 2:
 * Input: grid = [[1,2,3],[4,5,6]]
 * Output: 12
 * 
 * Constraints:
 * m == grid.length
 * n == grid[i].length
 * 1 <= m, n <= 200
 * 0 <= grid[i][j] <= 100
 * 
 *** =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-="""

 class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        if not grid:
        	return 0

        row, col = len(grid), len(grid[0])
        dp = [[0 for _ in range(col)] for _ in range(row)]
        dp[0][0] = grid[0][0]

        for index in range(1, row):
        	dp[index][0] = dp[index-1][0] + grid[index][0]

        for index in range(1, col):
        	dp[0][index] = dp[0][index-1] + grid[0][index]

        print dp
        for index_i in range(1, row):
        	for index_j in range(1, col):
        		dp[index_i][index_j] = min(dp[index_i-1][index_j], dp[index_i][index_j-1]) + grid[index_i][index_j]

        return dp[row-1][col-1]

File: PHP/0064_minimum-path-sum.php
<?php
/*=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 * 64. Minimum Path Sum
 *
 * Medium
 *
 * Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right,
 * which minimizes the sum of all numbers along its path.
 *
 * Note: You can only move either down or right at any point in time.
 *
 * Example 1:
 * Input: grid = [[1,3,1],[1,5,1],[4,2,1]]
 * Output: 7
 * Explanation: Because the path 1 → 3 → 1 → 1 → 1 minimizes the sum.
 *
 * Example 2:
 * Input: grid = [[1,2,3],[4,5,6]]
 * Output: 12
 *
 * Constraints:
 * m == grid.length
 * n == grid[i].length
 * 1 <= m, n <= 200
 * 0 <= grid[i][j] <= 100
 *
 *** =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=*/

 class Solution
 {
     /**
      * @param Integer[][] $grid
      * @return Integer
      */
     public function minPathSum($grid)
     {
         $m = count($grid);
         $n = count($grid[0]);
         $dp = array_fill(0, $m, array_fill(0, $n, 0));
         
         $dp[0][0] = $grid[0][0];

         for ($i = 1; $i < $m; $i++) {
             $dp[$i][0] = $dp[$i-1][0] + $grid[$i][0];
         }
         
         for ($i = 1; $i < $n; $i++) {
             $dp[0][$i] = $dp[0][$i-1] + $grid[0][$i];
         }
        
         for ($j = 1; $j < $n; $j++) {
             for ($i = 1; $i < $m; $i++) {
                 $dp[$i][$j] = $grid[$i][$j] + min($dp[$i-1][$j], $dp[$i][$j-1]);
             }
         }
         
         return $dp[$m-1][$n-1];
     }
 }
